File: HOT100/8.2/105.从前序与中序遍历序列构造二叉树.py
#
# @lc app=leetcode.cn id=105 lang=python3
#
# [105] 从前序与中序遍历序列构造二叉树
#
from mytools import *
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        
        def dfs(pl, pr, il, ir):
            if pl > pr: return None
            tmp = preorder[pl]
            node = TreeNode(tmp)
            if pl == pr: return node
            mid = -1
            for i in range(il, ir + 1):
                if inorder[i] == tmp:
                    mid = i
                    break
            pl += 1
            lil, lir, ril, rir = il, mid - 1, mid + 1,ir
            lpl, lpr, rpl, rpr = pl, pl + lir - lil, pl + lir - lil + 1, pr
            node.left = dfs(lpl, lpr, lil, lir)
            node.right = dfs(rpl, rpr, ril, rir)
            return node
        return dfs(0, len(preorder) - 1, 0, len(inorder) - 1)
# @lc code=end

File: HOT100p2/105.从前序与中序遍历序列构造二叉树.cpp
/*
 * @lc app=leetcode.cn id=105 lang=cpp
 *
 * [105] 从前序与中序遍历序列构造二叉树
 */
struct TreeNode
{
    int val;
    TreeNode *left, *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {};
};

// @lc code=start
#pragma once
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
#pragma GCC optimize("Ofast,no-stack-protector,unroll-loops")
#define ALL(v) v.begin(), v.end()
#define For(i, _) for (int i = 0, i##end = _; i < i##end; ++i)
#define FOR(i, _, __) for (int i = _, i##end = __; i < i##end; ++i)
#define Rep(i, _) for (int i = (_); i >= 0; --i)
#define REP(i, __, _) for (int i = (__), i##end = _; i >= i##end; --i)
typedef long long ll;
typedef unsigned long long ull;
#define V vector
#define pb push_back
#define pf push_front
#define qb pop_back
#define qf pop_front
#define eb emplace_back
typedef pair<int, int> pii;
typedef pair<ll, int> pli;
#define fi first
#define se second
const int dir[4][2] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
const int inf = 0x3f3f3f3f;
const ll infl = 0x3f3f3f3f3f3f3f3fll;
const int mod = 1e9 + 7;
template <class T>
inline bool ckmin(T &x, const T &y) { return x > y ? (x = y, true) : false; }
template <class T>
inline bool ckmax(T &x, const T &y) { return x < y ? (x = y, true) : false; }
int init = []()
{ ios::sync_with_stdio(false); cin.tie(nullptr); return 0; }();

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        auto dfs = [&] (auto&& dfs, int pl, int pr, int il, int ir) -> TreeNode* {
            if (pl > pr) return NULL;
            int tmp = preorder[pl];
            TreeNode *cur = new TreeNode(tmp);
            if (pl == pr) return cur;
            int mid = -1;
            FOR(i, il, ir + 1) if (inorder[i] == tmp) {
                mid = i;
                break;
            }
            int diff = mid - il;
            int lpl = pl + 1, lpr = lpl + diff - 1, rpl = lpr + 1, rpr = pr;
            int lil = il, lir = mid - 1, ril = mid + 1, rir = ir;
            cur->left = dfs(dfs, lpl, lpr, lil, lir);
            cur->right = dfs(dfs, rpl, rpr, ril, rir);
            return cur;
        };
        return dfs(dfs, 0, preorder.size() - 1, 0, inorder.size() - 1);
    }
};
// @lc code=end

